packages/@aws-cdk/aws-iam/lib/policy-statement.ts

```diff
@@ -408,6 +409,27 @@ export class PolicyStatement {
     }
     this.addConditions(conditions);
   }
+
+  /**
+   * Validate PolicyStatementProps
+   *
+   * As per [IAM policy grammar](https://docs.aws.amazon.com/en_us/IAM/latest/UserGuide/reference_policies_grammar.html#policies-grammar-bnf),
+   * the action and resource blocks are mandatory.
+   *
+   * @param props
+   */
+  private validateProps(props: PolicyStatementProps) {
+    if (props.conditions || props.effect || props.notPrincipals || props.notResources || props.principals || props.resources || props.sid) {
+      if (!props.actions && !props.notActions) {
+        throw new Error('Action block is mandatory. Either `actions` or `notActions` prop must be specified');
+      }
+    }
+    if (props.actions || props.conditions || props.effect || props.notActions || props.notPrincipals || props.principals || props.sid) {
+      if (!props.resources && !props.notResources) {
```

Comment:
I also have a question on [IAM policy grammar](https://docs.aws.amazon.com/en_us/IAM/latest/UserGuide/reference_policies_grammar.html#policies-grammar-bnf). It says the action and resource blocks are mandatory.  However, for certain policies (example: trust relationship policy) `resource` block isn't allowed. Here's the instance of such [case](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-cloudfront/lib/web_distribution.ts#L923):
```ts
 for (const a of input.lambdaFunctionAssociations) {
        if (a.lambdaFunction.role && a.lambdaFunction.role instanceof iam.Role && a.lambdaFunction.role.assumeRolePolicy) {
          a.lambdaFunction.role.assumeRolePolicy.addStatements(new iam.PolicyStatement({
            actions: [ 'sts:AssumeRole' ],
            principals: [ new iam.ServicePrincipal('edgelambda.amazonaws.com') ],
          }));
        }
      }
```

Please let me know if I'm missing something. I will move this PR to draft in the meantime.

Reply:
The docs aren't super clear about this, but there is a distinction between:

* Identity policies (what permissions an actor is allowed to perform, generally called IAM policies)
* Resource policies (what actions does the resource allow to be performed TO IT, you can think of these as "trust" policies although that is not a name that is in general use).

The syntax between these two is quite similar, except:

* An identity policy generally does not need to indicate the principal, as it is always about *this* principal.
* A resource policy generally does not need to indicate the resource, as it is always about *this* resource (of course there are exceptions: some resources like Queues *require* a `*`, and some resources like Buckets allow you to use ARNs to identify objects in the bucket).

Now here's the catch:

**An AssumeRolePolicy is actually a *resource* policy on an *identity* object**.

You can see that that's true, because the AssumeRolePolicy defines what actions are allowed to performed *on* the role (very specifically, who is allowed to Assume it).

And so therefore, it does not take a Resource. The resource is implicitly: this role.